File: project5/LSMC.R
Monomial = function (x, k){
  if (k == 1)
    return(x^0)
  if (k == 2)
    return(x)
  if (k == 3)
    return(x^2)
  if (k == 4)
    return(x^3)
}

Hermit = function (x, k){
  if (k == 1)
    return(x^0)
  if (k == 2)
    return(2*x)
  if (k == 3)
    return(4*x^2-2)
  if (k == 4)
    return(8*x^3-12*x)
}

Lagur = function (x, k){
  if (k == 1)
    return(exp(-x/2))
  if (k == 2)
    return(exp(-x/2)*(1-x))
  if (k == 3)
    return(exp(-x/2)*(1- 2*x+x^2/2))
  if (k == 4)
    return(exp(-x/2)*(1 - 3*x + 3*x^2/2 -x^3/6))
}

LSMC = function (N, t, K, S0, r, sigma, num, k, FUN){
  paths = matrix(rep(0, len = num *(N+1)), nrow = num)
  rand = matrix(rnorm(num*N), nrow = num)
  CF = matrix(rep(0, len = num *(N)), nrow = num)
  dt = t/N
  paths[, 1] = S0
  zeros = rep(0 , len = num)
  zeros_row = rep(0, len = N)
  for (i in 1:N){
    paths[, i+1] = paths[, i] *exp((r-sigma^2/2)*dt + sigma*rand[,i]*sqrt(dt))
  }
  CF[, N] = apply(cbind(K - paths[,N+1], zeros), 1, max)
  for (i in (N-1) : 1){
    temp = (K - paths[,i+1] > 0)
    x = paths[, i+1] * temp
    dis_factor = exp(seq(-r*dt,-(N-i)*r*dt, -r*dt))
    if (i == N-1){
      y = CF[, (i+1):N] * dis_factor * temp
    }else{
      y = CF[, (i+1):N] %*% dis_factor * temp
    }
    pos = seq(1, num, 1) * temp
    y_clean = y[pos]
    x_clean = x[pos]
    A = matrix(rep(0, len = k*k), nrow = k)
    B = matrix(rep(0, len =k), nrow = k)
    X = c()
    for (a in 1:k){
      for (b in 1:k){
        A[a,b] = sum(FUN(x_clean, a) * FUN(x_clean, b))
      }
      B[a,1] = sum(FUN(x_clean,a) * y_clean)
      X = cbind(X, FUN(x, a))
    }
    reg = solve(A,B)
    continuation = X %*% reg * temp
    newCF = apply(cbind(K - paths[,i+1], zeros), 1, max)

    for (j in 1: num){
      if (newCF[j] > continuation[j]){
        CF[j, ] = zeros_row
        CF[j,i] = newCF[j]
      }
    }
  }
  D = exp(seq(-r*dt,-N*r*dt, -r*dt))
  return(mean(CF%*% D))
}


N = 250
S0 = 36
r = 0.06
sigma = 0.2
K = 40
t = 1
num = 100000
k = 3
#out = LSMC(N, t, K, S0, r, sigma, num, k, Monomial)
#out2 = LSMC(N, t, K, S0, r, sigma, num, k, Hermit)
out3 = LSMC(N, t, K, S0, r, sigma, num, k, Lagur)

File: project5/LSMC.m
function [Hermit, mono, Lagur] = LSMC(N, T, K, S0, r, sigma, num, k)
    rng(12345);
    paths = zeros(num, N+1);
    rnorm = randn(num, N);
    CF = zeros(num, N);
    CF2 = zeros(num, N);
    CF3 = zeros(num, N);
    dt = T/N;
    paths(:, 1) = S0;
    for i = 1:N
        paths(:,i+1) = paths(:, i).*exp( (r-sigma^2/2)*dt +  sigma* rnorm(:, i) * sqrt(dt));
    end
    CF(:, N) = max([K - paths(:, N+1) zeros(num,1)],[],2);
    CF2(:, N) = max([K - paths(:, N+1) zeros(num,1)],[],2);
    CF3(:, N) = max([K - paths(:, N+1) zeros(num,1)],[],2);
    for i = (N-1):-1:1
        temp = (paths(:,i+1) < K);
        x = paths(:, i+1).* temp;
        y = zeros(num, 1);
        y2 = zeros(num, 1);
        y3 = zeros(num, 1);
        for j = 1: num
            [M,I] = max(CF(j, i+1:N));
            [M2,I2] = max(CF2(j, i+1:N));
            [M3,I3] = max(CF3(j, i+1:N));
            if M > 0
                y(j) = CF(j, i+I) * exp(-r*dt*I);
            end
            if M2 > 0
                y2(j) = CF2(j, i+I2) * exp(-r*dt*I2);
            end
            if M3 > 0
                y3(j) = CF3(j, i+I3) * exp(-r*dt*I3);
            end
        end
        y = y.* temp;
        y2 = y2.* temp;
        y3 = y3.* temp;
        if k == 2
            X = [ones(length(x),1).* temp x];
            X2 = [ones(length(x),1).* temp 2*x];
            X3 = [exp(-x/2).* temp exp(-x/2).*(1-x).* temp];
        elseif k == 3
            X = [ones(length(x),1).* temp x x.^2];
            X2 = [ones(length(x),1).* temp 2*x 4*x.^2 - 2];
            X3 = [exp(-x/2).* temp exp(-x/2).*(1-x).* temp exp(-x/2).*(1-2*x+x.^2/2).* temp];
        elseif k == 4
            X = [ones(length(x),1).* temp x x.^2 x.^3];
            X2 = [ones(length(x),1).* temp 2*x 4*x.^2 - 2 8*x.^3 - 12*x];
            X3 = [exp(-x/2).* temp exp(-x/2).*(1-x).* temp exp(-x/2).*(1-2*x+x.^2/2).* temp exp(-x/2).*(1-3*x+3*x.^2/2 - x.^3/6).* temp];
        end
        b = X\y;
        b2 = X2\y2;
        b3 = X3\y3;
        continuoation = X*b;
        continuoation2 = X2*b2;
        continuoation3 = X3*b3;
        newCF = max([K - paths(:,i+1) zeros(num, 1)], [], 2);
        ngc = (continuoation < newCF);
        ngc2 = (continuoation2 < newCF);
        ngc3 = (continuoation3 < newCF);
        for j = 1: length(ngc)
            if ngc(j) == 1
                CF(j, : ) = zeros(1, N);
                CF(j, i) = newCF(j);
            end    
            if ngc2(j) == 1
                CF2(j, : ) = zeros(1, N);
                CF2(j, i) = newCF(j);
            end  
            if ngc3(j) == 1
                CF3(j, : ) = zeros(1, N);
                CF3(j, i) = newCF(j);
            end  
        end
    end
    
    B = transpose(exp(-r*dt:-r*dt:-N*r*dt));
    mono = mean(CF * B);
    Hermit = mean(CF2 * B);
    Lagur = mean(CF3 * B);

end